src/compiler/checker.ts

```diff
@@ -7156,6 +7156,25 @@ namespace ts {
                 }
             }
 
+            function isUnionOrIntersectionTypeWithoutNullableConstituents(type: Type): boolean {
+                if (!(type.flags & TypeFlags.UnionOrIntersection)) {
+                    return false;
+                }
+                // at this point we know that this is union or intersection type possibly with nullable constituents.
+                // check if we still will have compound type if we ignore nullable components.
+                let seenNonNullable = false;
+                for (const t of (<UnionOrIntersectionType>type).types) {

```

Comment:
no, since `maybeTypeOfKind(type, ~TypeFlags.Nullable)` will return `true` for union or intersection type that consist of non-nullable type and nullable type and this is not what we want.